### PR TITLE
Support using container tag in rbe_autoconfig().

### PR DESCRIPTION
**Pull request by xingao267**

### CODE REVIEW

File: rules/rbe_repo.bzl [hunk 1]
```diff
@@ -315,26 +317,11 @@ def _validate_host(ctx):
...
 
-# Checks if a standard published config can be used for the given revision
-# of the rbe ubuntu 16_04 container. If so, produces BUILD files with aliases
-# for all the required toolchain / platform targets. Otherwise returns false.
-def _use_standard_config(ctx, bazel_version, bazel_rc_version, revision):
-    # If rc version is required, simply return
-    if bazel_rc_version:
-        return False
+# Produces BUILD files with aliases for all the required toolchain / platform targets.
+def _use_standard_config(ctx, bazel_version, bazel_rc_version, config_version):
```
Comment (nlopezgi) [inline rules/rbe_repo.bzl]: bazel_rc_version is no longer used in this def

File: rules/rbe_repo.bzl [hunk 2]
```diff
@@ -718,91 +702,169 @@ def rbe_autoconfig(
...
+        tag,
+        digest,
+        use_checked_in_confs,
+        env,
+        java_home,
+        bazel_version,
+        bazel_rc_version,
+    )
+
+    use_checked_in_confs = False
```
Comment (nlopezgi) [inline rules/rbe_repo.bzl]: why are you setting this here to false?

File: rules/rbe_repo.bzl [hunk 3]
```diff
@@ -718,91 +702,169 @@ def rbe_autoconfig(
...
+    digest = digest
+    tag = tag
+    config_version = None
+    if resolved_config_info:
+        use_checked_in_confs = True
```
Comment (nlopezgi) [inline rules/rbe_repo.bzl]: why do you need to override use_checked_in_confs here?

File: rules/rbe_repo.bzl [hunk 4]
```diff
@@ -718,91 +702,169 @@ def rbe_autoconfig(
...
+        digest,
+        use_checked_in_confs,
+        env,
+        java_home,
+        bazel_version,
+        bazel_rc_version,
+    )
+
+    use_checked_in_confs = False
+    digest = digest
```
Comment (nlopezgi) [inline rules/rbe_repo.bzl]: what's the purpose of line 790 and 791?

File: rules/rbe_repo.bzl [hunk 5]
```diff
@@ -718,91 +702,169 @@ def rbe_autoconfig(
...
         env = env,
         exec_compatible_with = exec_compatible_with,
         java_home = java_home,
         output_base = output_base,
         registry = registry,
         repository = repository,
-        revision = revision,
+        tag = tag,
         target_compatible_with = target_compatible_with,
         use_checked_in_confs = use_checked_in_confs,
```
Comment (nlopezgi) [inline rules/rbe_repo.bzl]: I think you dont need to pass down use_checked_in_confs and you can just check if config_version is not None

File: rules/rbe_repo.bzl [hunk 6]
```diff
@@ -718,91 +702,169 @@ def rbe_autoconfig(
...
+            tag = RBE_UBUNTU16_04_LATEST
+        digest = public_rbe_ubuntu16_04_sha256s().get(tag, None)
+        if not digest:  # We didn't find checked-in configs in this repo.
+            return None
+    else:  # Implies tag is not specified
+        for pair in public_rbe_ubuntu16_04_sha256s().items():
+            if pair[1] == digest:
+                tag = pair[0]
+                break
+        if not tag:
```
Comment (nlopezgi) [inline rules/rbe_repo.bzl]: document which cases this would happen

File: rules/rbe_repo.bzl [hunk 7]
```diff
@@ -718,91 +702,169 @@ def rbe_autoconfig(
...
+            return None
+
+    # Verify a toolchain config exists for the given version of Bazel and the
+    # given tag of the container
+    config_version = public_rbe_ubuntu16_04_config_version().get(tag, None)
+    if not config_version:
```
Comment (nlopezgi) [inline rules/rbe_repo.bzl]: do 860 and 863 in a if block

File: rules/rbe_repo.bzl (hunk 5 above)
Comment (xingao267) [reply rules/rbe_repo.bzl]: Done.

File: rules/rbe_repo.bzl (hunk 3 above)
Comment (xingao267) [reply rules/rbe_repo.bzl]: Removed.

File: rules/rbe_repo.bzl (hunk 4 above)
Comment (xingao267) [reply rules/rbe_repo.bzl]: Removed.

File: rules/rbe_repo.bzl (hunk 2 above)
Comment (xingao267) [reply rules/rbe_repo.bzl]: Removed.

File: rules/rbe_repo.bzl (hunk 6 above)
Comment (xingao267) [reply rules/rbe_repo.bzl]: Done.

File: rules/rbe_repo.bzl (hunk 7 above)
Comment (xingao267) [reply rules/rbe_repo.bzl]: Done.

File: rules/rbe_repo.bzl [hunk 8]
```diff
@@ -611,21 +600,19 @@ _rbe_autoconfig = repository_rule(
...
-                   "is non-release version."),
-        ),
         "config_dir": attr.string(
             doc = ("Optional. Use only if output_base is defined. If you want to " +
                    "create multiple toolchain configs (for the same version of Bazel) " +
                    "you can use this attr to indicate a type of config (e.g., default, " +
                    "msan). The configs will be generated in a sub-directory when this attr  " +
                    "is used."),
         ),
+        "config_version": attr.string(
```
[the remark this answers is not in the document]
Comment (xingao267) [reply rules/rbe_repo.bzl]: Done

File: rules/rbe_repo.bzl [hunk 9]
```diff
@@ -611,21 +600,19 @@ _rbe_autoconfig = repository_rule(
...
+            doc = ("The config version found for the given container and Bazel version." +
+                   "Used internally when use_checked_in_confs is true."),
+        ),
         "digest": attr.string(
-            mandatory = True,
-            doc = ("The digest (sha256 sum) of the image to pull. For example, " +
+            doc = ("Optional. The digest (sha256 sum) of the image to pull. For example, " +
```
[the remark this answers is not in the document]
Comment (xingao267) [reply rules/rbe_repo.bzl]: Done

File: rules/rbe_repo.bzl [hunk 10]
```diff
@@ -718,91 +700,151 @@ def rbe_autoconfig(
           `Use only (major, minor, patch), e.g., '0.20.0'. Default is "local"
           which means the same version of Bazel that is currently running will
           be used. If local is a non release version, rbe_autoconfig will fallback
-          to using the latest release version (see default for bazel_version_fallback
-          in attrs of _rbe_autoconfig for current latest).
-      bazel_rc: The rc (for the given version of Bazel) to use. Must be published
-          in https://releases.bazel.build
+          to using the latest release version (see _BAZEL_VERSION_FALLBACK).
+      bazel_rc_version: The rc (for the given version of Bazel) to use. Must be published
```
[the remark this answers is not in the document]
Comment (xingao267) [reply rules/rbe_repo.bzl]: Done

File: rules/rbe_repo.bzl [hunk 11]
```diff
@@ -718,91 +700,151 @@ def rbe_autoconfig(
...
+             "all of 'tag', 'repository' and 'registry' or " +
+             "none of them must be set.")
+
+    # Set to defaults only if all are unset.
+    if not repository and not registry and not tag and not digest:
+        repository = _RBE_UBUNTU_REPO
+        registry = _RBE_UBUNTU_REGISTRY
+        tag = RBE_UBUNTU16_04_LATEST
+
+    if registry and registry == _RBE_UBUNTU_REGISTRY and repository and repository == _RBE_UBUNTU_REPO and not env:
```
[the remark this answers is not in the document]
Comment (xingao267) [reply rules/rbe_repo.bzl]: Done

File: rules/rbe_repo.bzl [hunk 12]
```diff
@@ -718,91 +700,151 @@ def rbe_autoconfig(
...
+        env = clang_env()
+
+    config_version = validateUseOfCheckedInConfigs(
+        registry,
```
[the remark this answers is not in the document]
Comment (xingao267) [reply rules/rbe_repo.bzl]: Done

File: rules/rbe_repo.bzl [hunk 13]
```diff
@@ -718,91 +700,151 @@ def rbe_autoconfig(
...
+        digest = public_rbe_ubuntu16_04_sha256s().get(tag, None)
+        if not digest:  # We didn't find checked-in configs in this repo.
+            return None
+    else:  # Implies tag is not specified
+        for pair in public_rbe_ubuntu16_04_sha256s().items():
+            if pair[1] == digest:
+                tag = pair[0]
+                break
+        if not tag:
+            # The given RBE Ubuntu1604 container digest is not one of the released ones.
```
[the remark this answers is not in the document]
Comment (xingao267) [reply rules/rbe_repo.bzl]: Done

**Pull request merged**